src/data-model-providers/codedriven/CodeDrivenDataModelProvider.cpp

```diff
@@ -0,0 +1,337 @@
+/*
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/server-cluster/ServerClusterContext.h>
+#include <data-model-providers/codedriven/CodeDrivenDataModelProvider.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <protocols/interaction_model/StatusCode.h>
+
+using chip::Protocols::InteractionModel::Status;
+namespace chip {
+namespace app {
+
+CHIP_ERROR CodeDrivenDataModelProvider::Startup(DataModel::InteractionModelContext context)
+{
+    ReturnErrorOnFailure(DataModel::Provider::Startup(context));
+
+    mServerClusterContext.reset();
+    mServerClusterContext.emplace(ServerClusterContext({
+        .provider           = this,
+        .storage            = mPersistentStorageDelegate,
+        .interactionContext = &mContext,
+    }));
+
+    // Startup all server clusters across all endpoints
+    bool had_failure = false;
+    for (auto * endpointProvider : mEndpointProviderRegistry)
+    {
+        ReadOnlyBufferBuilder<ServerClusterInterface *> serverClusterBuilder;
+        ReturnErrorOnFailure(endpointProvider->ServerClusterInterfaces(serverClusterBuilder));
+        auto serverClusters = serverClusterBuilder.TakeBuffer();
+
+        for (auto * serverCluster : serverClusters)
+
+        {
+            CHIP_ERROR err = serverCluster->Startup(*mServerClusterContext);
+            if (err != CHIP_NO_ERROR)
+            {
+                had_failure = true;
+                VerifyOrReturnError(!serverCluster->GetPaths().empty(), CHIP_ERROR_INCORRECT_STATE);
```

Comment:
if paths is empty, we would return a incorrect state and lose the startup error. Could we log the startup error with some ("no paths available") instead and not consider this a critical failure (since we do not consider startup failure critical ... so why would a path failure be?)